ansible_events_ui/ruleset.py

```diff
@@ -27,18 +27,31 @@
 import tempfile
 from functools import partial
 
+import aiodocker
 import ansible_runner
-from sqlalchemy import insert, select
+from sqlalchemy import insert
 from sqlalchemy.ext.asyncio import AsyncSession
 
-from .db.models import job_instance_events, playbooks
+from ansible_events_ui.managers import taskmanager
+
+from .config import load_settings
+from .db.models import activation_instance_logs, job_instance_events
 from .managers import updatemanager
 from .messages import JobEnd
 
 logger = logging.getLogger("ansible_events_ui")
 
 activated_rulesets = {}
 ansible_events = shutil.which("ansible-events")
+ssh_agent = shutil.which("ssh-agent")
```

Comment:
Can you please explain why do you call `shutil.which` for each CLI tool the application uses.

Reply:
subprocess calls should use exact paths to binaries and use shell=False to prevent code injection.

Reply:
FYI `subprocess` can resolve executables in `PATH` even when called with `shell=False`, because it relies on `os.execvpe()` system call, but I double checked and it appears there is a recommendation added recently in python&nbsp;3.10&nbsp;[1] that recommends using `shutil.which` for maximum reliability, so it makes sense to me.

[1] https://docs.python.org/3.10/library/subprocess.html#subprocess.Popen